### DISCUSSION TITLE
How to kill jobs by there job-id instead of there pid

### DISCUSSION THREAD

**Discussion by FranzAtGithub** [xonsh/xonsh#4524] [Q&A] [answered]
Hi,
I have used bash for many years now but I am considering xonsh as a replacement.
For far it is working quite nicely for me but one thing is bothering me quite a bit.
In bash I was used to kill background jobs by there job id (mostly a single digit; easily typed).
In xonsh that does not work for me. The `fg` and `bg` commands work on job ids but kill does not.

Aything wrong in my setup?

**Comment by FranzAtGithub** [comment]
Hi,
I wonder why I did not get any response to my question.

Is my question unclear or is the desired behavior simply not available in xonsh?
Or maybe this was answered somewhere else already and I just didn't find it?
Please let me know if I can provide any more information or if I should rephrase my question more clearly.

**Reply by gforsyth** [reply]
Hey @FranzAtGithub -- sorry I missed this the first time around.
It seems like we're missing this bit of functionality but it shouldn't be too hard to implement.  You can `disown` jobs by their assigned (non-pid) identifier, but you can't kill them.

**Comment by FranzAtGithub** [comment]
Ok so this will be possible at some point I guess.

I tried to check the code myself though I have very limited experience with larger projects.
From want I get. A job is simply a `dict` that has among others a `pids` key (could it be more than just one?).

From https://github.com/xonsh/xonsh/blob/c2d25ac2519fa4edb21f2c8e155495963ad8a27d/xonsh/jobs.py#L261-L262 it seems that the words "task" and "job" are used interchangeably and what I call job-id is usually the `tid` value.

On https://github.com/xonsh/xonsh/blob/c2d25ac2519fa4edb21f2c8e155495963ad8a27d/xonsh/jobs.py#L117 `_kill` takes a `job` argument. Should this be modified such that it could also handle an `tid: int`? How could both cases be distinguished?

**Reply by FranzAtGithub** [reply]
Sorry this made zero sense. Somehow the `kill` command already gets a PID. So somewhere else this PID is already mapped to the correct job.

**Reply by gforsyth** [reply]
You could do something like this in your `xonshrc` as a temporary fix:
```xonsh
from xonsh.jobs import _kill

def _xkill(args):
    taskid = int(args[0])
    job = __xonsh__.all_jobs.get(taskid)
    if job is None:
        raise XonshError(f"kill: sending signal to task-id: {taskid} failed: No such process")
    _kill(job)

aliases["xkill"] = _xkill
```

and then you can do this:

```xonsh
(base) gil@badcat ~ 🐚  jobs
[1]+ stopped: vim (43372)
(base) gil@badcat ~ 🐚  xkill 1
(base) gil@badcat ~ 🐚  jobs
```

If you wanted to roll something like this into `xonsh` itself, it would probably be best to have it be part of a subcommand of the `jobs` builtin, so you could run `jobs kill <task-id>` (another option would be to try to overload `kill` but I think that would lead to problems)

**Reply by FranzAtGithub** [reply]
Ty. Works for me.

I would have liked to play around and see if I can come up with a PR. But I cannot even figure out the order of function calls ^^
Maybe having never implemented even a simple shell is also not a ideal.

PS: Do you want to make your comment an answer actual answer?

**Comment by gforsyth** [comment, marked as answer]
Copying from comment below:

You could do something like this in your `xonshrc` as a temporary fix:

```
from xonsh.jobs import _kill

def _xkill(args):
    taskid = int(args[0])
    job = __xonsh__.all_jobs.get(taskid)
    if job is None:
        raise XonshError(f"kill: sending signal to task-id: {taskid} failed: No such process")
    _kill(job)

aliases["xkill"] = _xkill
```

and then you can do this:

```
(base) gil@badcat ~ 🐚  jobs
[1]+ stopped: vim (43372)
(base) gil@badcat ~ 🐚  xkill 1
(base) gil@badcat ~ 🐚  jobs
```

If you wanted to roll something like this into `xonsh` itself, it would probably be best to have it be part of a subcommand of the `jobs` builtin, so you could run `jobs kill <task-id>` (another option would be to try to overload kill but I think that would lead to problems)
